File: 57-insert-interval/insert-interval.js
var insert = function(intervals, newInterval) {
    let i = 0;
    let n = intervals.length;
    let result = [];
    
    while (i < intervals.length) {
        if (intervals[i][1] < newInterval[0])
            result.push(intervals[i]);
        else if (intervals[i][0] > newInterval[1]){
            break;
        } else {
            //Overlap : merge them
            newInterval[0] = Math.min(newInterval[0], intervals[i][0]);
            newInterval[1] = Math.max(newInterval[1], intervals[i][1]);
        }
        i++;
    }
    
    result.push(newInterval);
    while (i < n){
        result.push(intervals[i++]);
    }
    
    return result;
};

File: 57-insert-interval/insert-interval.kt
class Solution {
    fun insert(intervals: Array<IntArray>, newInterval: IntArray): Array<IntArray> {
        var i = 0
        val n = intervals.size
        val result = mutableListOf<IntArray>()
        
        while (i < intervals.size) {
            if (intervals[i][1] < newInterval[0])
                result.add(intervals[i])
            else if (intervals[i][0] > newInterval[1]){
                break
            } else {
                //Overlap : merge them
                newInterval[0] = minOf(newInterval[0], intervals[i][0])
                newInterval[1] = maxOf(newInterval[1], intervals[i][1])
            }
            i++
        }
        
        result.add(newInterval)
        while (i < n){
            result.add(intervals[i++])
        }
        
        return result.toTypedArray()
    }
}